tests/e2e/cleanup_test.go

```diff
@@ -25,53 +25,127 @@ import (
+
+	cleanupNode(ctx, clientSet)
+	cleanupNFDObjects(ctx, nfdClient, testNamespace.Name)
+	cleanupCRDs()
+}
+
+// waitForDeletion polls the provided checkFunc until a NotFound error is returned,
+// confirming that the resource is deleted.
+func waitForDeletion(resourceName string, checkFunc func() error) error {
+	timeout := 2 * time.Minute
+	interval := 5 * time.Second
+	start := time.Now()
+	for {
+		err := checkFunc()
+		if err != nil && errors.IsNotFound(err) {
+			return nil
+		}
+		if time.Since(start) > timeout {
+			return fmt.Errorf("timed out waiting for deletion of %s", resourceName)
+		}
+		time.Sleep(interval)
 	}
+}
 
-	controlPlaneTaint := corev1.Taint{
-		Effect: corev1.TaintEffectNoSchedule,
-		Key:    "node-role.kubernetes.io/control-plane",
+// cleanupTestPods deletes all test Pods in the namespace that have the label "app.nvidia.com=k8s-dra-driver-gpu-test-app".
+func cleanupTestPods(namespace string) error {
+	labelSelector := "app.nvidia.com=k8s-device-plugin-test-app"
+	podList, err := clientSet.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions{
+		LabelSelector: labelSelector,
+	})
+	if err != nil {
+		return err
 	}
-	out := []corev1.Node{}
-	for _, node := range nodeList.Items {
-		if !TaintExists(node.Spec.Taints, &controlPlaneTaint) {
-			out = append(out, node)
+
+	zero := int64(0)
+	deleteOptions := metav1.DeleteOptions{GracePeriodSeconds: &zero}
+	for _, pod := range podList.Items {
+		if err = clientSet.CoreV1().Pods(namespace).Delete(ctx, pod.Name, deleteOptions); err != nil {
+			return err
+		}
+		if err = waitForDeletion(pod.Name, func() error {
+			_, err := clientSet.CoreV1().Pods(namespace).Get(ctx, pod.Name, metav1.GetOptions{})
+			return err
+		}); err != nil {
+			return err
 		}
 	}
+	return nil
+}
 
-	if len(out) == 0 {
-		return nil, fmt.Errorf("no non-control-plane nodes found in the cluster")
+// cleanupHelmDeployments uninstalls all deployed Helm releases in the specified namespace.
+func cleanupHelmDeployments(namespace string) error {
+	releases, err := helmClient.ListDeployedReleases()
+	if err != nil {
+		return fmt.Errorf("failed to list deployed releases: %w", err)
 	}
-	return out, nil
+
+	for _, release := range releases {
+		// Check if the release is deployed in the target namespace.
+		// Depending on your helmClient configuration the release might carry the namespace information.
+		if release.Namespace == namespace {
+			if err := helmClient.UninstallReleaseByName(release.Name); err != nil {
+				return fmt.Errorf("failed to uninstall release %q: %w", release.Name, err)
+			}
+		}
+	}
+	return nil
 }
 
-func GetNode(nodes []corev1.Node, nodeName string) corev1.Node {
-	for _, node := range nodes {
-		if node.Name == nodeName {
-			return node
+// deleteTestNamespace deletes the test namespace and waits for its deletion.
+func deleteTestNamespace() {
+	defer func() {
+		err := clientSet.CoreV1().Namespaces().Delete(ctx, testNamespace.Name, metav1.DeleteOptions{})
```

Comment:
Here as well.

methods like cleanupNode and cleanupNFDObjects have the `ctx` parameter

Reply:
same as above